### PR TITLE
Add workflow collections

### LINKED ISSUE #2550: Get all Collection items
Re: https://github.com/OpenFn/lightning/issues/2190

Create a Collections module that exposes a function that returns all values for a given collection name.

The result should be a stream of structs containing:

- `key` : string
- `value` : map

This is likely the first issue in the epic above, and so we will need some tables.

- `collections`
    - `name` (unique)
    - belongs to `project`
    - timestamps
- `collection_items`
    - `key` 
    - belongs to `project`

### Acceptance criteria

- Migrations for Collection tables
- A new context module
- A function in the module that returns a stream of items

### LINKED ISSUE #2551: Get Collection item by key
Re: https://github.com/OpenFn/lightning/issues/2190

Add a function that retrieves a single item from a collection based on a key.

### Acceptance criteria

- A new function in the Collections context module

### LINKED ISSUE #2552: Get Collection by wildcards
Re: https://github.com/OpenFn/lightning/issues/2190

Add a function that retrieves a list of items from a collection based on a key with wildcards.

Given this rough example:
```js
{op: "get", collection: "collection-name", key: "foo:*:bar*"}
```
The resulting query might look something like this:
```sql
SELECT * FROM collection_items 
  -- join collections by name
  WHERE key LIKE "foo:%:bar%"
```

> Note that a simple BTREE index will help up until the first wildcard appears, using a GIN index with the pg_trgm extension will allow for indexing to work with more than one wildcard:
> 
> `CREATE INDEX idx_my_keys_trgm ON my_keys USING GIN (key_name gin_trgm_ops);`

### Acceptance criteria

- A new function in the Collections context that returns a stream of items with keys matching the query.
- Must be injection proof, i.e. use the `like` macros from Ecto.

### PR DESCRIPTION
**Pull request by jyeshe**
### Description

This PR adds workflow collections which are scoped to projects and provide similar functionality to Redis HSET and HGET (named hashes operations). It will be used not only to share mutable data between jobs on downstream but also shared among multiple workflows and it will be the responsibility of the programmer to use it without conflicts between workflows.

EDIT: The PRs are "merged" and it also includes functionalities similar to HGETALL and HSCAN allowing a query by pattern using a cursor (and respecting a limit).

Closes #2550 
Closes #2551
Closes #2552 

### Validation steps

There is not UI using it yet but a complete test coverage is provided.

## AI Usage

Please disclose how you've used AI in this work (it's cool, we just want to know!):

- [ ] Code generation (copilot but not intellisense)
- [ ] Learning or fact checking
- [ ] Strategy / design
- [ ] Optimisation / refactoring
- [ ] Translation / spellchecking / doc gen
- [ ] Other
- [x] I have not used AI

You can read more details in our [Responsible AI Policy](https://www.openfn.org/ai#pull-request-templates)

### Pre-submission checklist

- [x] I have performed a **self-review** of my code.
- [x] I have implemented and tested all related **authorization policies**. (e.g., `:owner`, `:admin`, `:editor`, `:viewer`)
- [ ] I have updated the **changelog**.
- [x] I have ticked a box in "AI usage" in this PR

### CODE REVIEW

File: priv/repo/migrations/20241009021209_create_collections.exs [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+    create table(:collections_items, primary_key: false) do
+      add :id, :binary_id, primary_key: true
+      add :key, :string
+      add :value, :string
+
+      add :collection_id,
+          references(:collections, on_delete: :delete_all, type: :binary_id, null: false)
+
+      timestamps()
```
Comment (stuartc) [inline priv/repo/migrations/20241009021209_create_collections.exs:25]: Could you change this to `timestamps(type: :naive_datetime_usec)` and the models as well.

Probably a good idea to start with sub-second resolution on these.

File: priv/repo/migrations/20241009021209_create_collections.exs [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    end
+
+    create unique_index(:collections_items, [:collection_id, :key])
```
Comment (stuartc) [inline priv/repo/migrations/20241009021209_create_collections.exs:28]: Nice one, do you think we should put an index on updated_at and inserted_at as well?

**Review by stuartc — CHANGES_REQUESTED** [review]

**Comment by josephjclark** [comment]
Thank you for the AI disclosure @jyeshe

File: priv/repo/migrations/20241009021209_create_collections.exs (hunk 1 above)
Comment (jyeshe) [reply priv/repo/migrations/20241009021209_create_collections.exs]: Done

**Review by stuartc — APPROVED** [review]

**Pull request merged**